pkg/controller/clusterrelocate/clusterrelocate_controller.go

```diff
@@ -0,0 +1,812 @@
+
+	labelSelector, err := metav1.LabelSelectorAsSelector(&clusterRelocate.Spec.ClusterDeploymentSelector)
+	if err != nil {
+		r.logger.WithError(err).
+			WithField("clusterRelocate", clusterRelocate.Name).
+			Warn("cannot parse clusterdeployment selector")
+		return
+	}
+
+	clusterDeployments := &hivev1.ClusterDeploymentList{}
+	if err := r.List(context.Background(), clusterDeployments); err != nil {
+		r.logger.WithError(err).
+			WithField("clusterRelocate", clusterRelocate.Name).
+			Log(controllerutils.LogLevel(err), "failed to list clusterdeployments")
+		return
+	}
+
+	for _, cd := range clusterDeployments.Items {
+		if relocateName, _, _ := controllerutils.IsRelocating(&cd); relocateName != clusterRelocate.Name &&
+			!labelSelector.Matches(labels.Set(cd.Labels)) {
+			continue
+		}
+		requests = append(requests,
+			reconcile.Request{NamespacedName: types.NamespacedName{Name: cd.Name, Namespace: cd.Namespace}})
+	}
+
+	return
+}
+
+// ReconcileClusterRelocate is the reconciler for ClusterRelocate. It will sync on ClusterDeployment resources
+// and relocate those that match with a ClusterRelocate.
+type ReconcileClusterRelocate struct {
+	client.Client
+	logger log.FieldLogger
+
+	// remoteClusterAPIClientBuilder is a function pointer to the function that gets a builder for building a client
+	// for the remote cluster's API server
+	remoteClusterAPIClientBuilder func(secret *corev1.Secret) remoteclient.Builder
+}
+
+// Reconcile relocates ClusterDeployments matching with a ClusterRelocate to another Hive instance.
+func (r *ReconcileClusterRelocate) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	start := time.Now()
+	logger := r.logger.WithFields(log.Fields{
+		"controller":        ControllerName,
+		"clusterDeployment": request.NamespacedName.String(),
+	})
+
+	// For logging, we need to see when the reconciliation loop starts and ends.
+	logger.Info("reconciling cluster deployment")
+	defer func() {
+		dur := time.Since(start)
+		hivemetrics.MetricControllerReconcileTime.WithLabelValues(ControllerName).Observe(dur.Seconds())
+		logger.WithField("elapsed", dur).Info("reconcile complete")
+	}()
+
+	// Fetch the ClusterDeployment instance
+	cd := &hivev1.ClusterDeployment{}
+	err := r.Get(context.TODO(), request.NamespacedName, cd)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Debug("cluster deployment not found")
+			return reconcile.Result{}, nil
+		}
+		logger.WithError(err).Error("Error getting cluster deployment")
+		return reconcile.Result{}, err
+	}
+
+	if cd.DeletionTimestamp != nil {
+		logger.Debug("skipping deleted clusterdeployment")
+		return reconcile.Result{}, err
+	}
+
+	currentRelocateName, relocateStatus, err := controllerutils.IsRelocating(cd)
+	if err != nil {
+		logger.WithError(err).Error("could not determine relocate status")
+		return reconcile.Result{}, errors.Wrap(err, "could not determine relocate status")
+	}
+
+	// Complete the relocation on the destination side
```

Comment:
Please clarify here the logic, because the cluster deployment is the last piece, and we're recocniling on a ClusterDeployment marked incoming, thus the destination side, and thus must be complete.